crates/ruff_linter/src/rules/flake8_implicit_str_concat/rules/implicit.rs

```diff
@@ -183,3 +188,41 @@ fn concatenate_strings(a_range: TextRange, b_range: TextRange, locator: &Locator
         range,
     )))
 }
+
+/// Pads an octal at the end of the string
+/// to three digits, if necessary.
+fn normalize_ending_octal(text: &str) -> String {
+    // Early return for short strings
+    if text.len() < 2 {
+        return text.to_string();
+    }
+
+    let mut rev_bytes = text.bytes().rev();
+    if let Some(last_byte @ b'0'..=b'7') = rev_bytes.next() {
+        // "\y" -> "\00y"
+        if has_odd_consecutive_backslashes(&rev_bytes) {
+            let prefix = &text[..text.len() - 2];
+            return format!("{prefix}\\00{}", last_byte as char);
+        }
+        // "\xy" -> "\0xy"
+        if let Some(penultimate_byte @ b'0'..=b'7') = rev_bytes.next() {
+            if has_odd_consecutive_backslashes(&rev_bytes) {
+                let prefix = &text[..text.len() - 3];
+                return format!(
+                    "{prefix}\\0{}{}",
+                    penultimate_byte as char, last_byte as char
+                );
+            }
+        }
+    }
+    text.to_string()
+}
+
+fn has_odd_consecutive_backslashes<I: Iterator<Item = u8> + Clone>(itr: &I) -> bool {
+    let mut itrclone = itr.clone();
+    let mut odd_backslashes = false;
+    while let Some(b'\\') = itrclone.next() {
+        odd_backslashes = !odd_backslashes;
+    }
+    odd_backslashes
+}
```

Comment:
nit: I'd probably rewrite the function like this:

```suggestion
fn has_odd_consecutive_backslashes(mut itr: impl Iterator<Item = u8>) -> bool {
    let mut odd_backslashes: bool = false;
    while let Some(b'\\') = itr.next() {
        odd_backslashes = !odd_backslashes;
    }
    odd_backslashes
}
```

And then explicitly call `rev_bytes.clone()` at each callsite before passing the iterator into this function. That makes the signature of this function slightly less complicated, and it also makes the code more explicit about each clone that we're doing. (The clone shouldn't be _that_ expensive here for an iterator instance, but it's still good to be explicit about it!)

Reply:
Good suggestion! implemented